### ISSUE TITLE
(Words IRC) Sending messages with tags

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6667]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ElementalAlchemist reported|
|-|-|
|Trac ID|trac#6667|
|Type|enhancement|
|Created|2013-08-06 03:02:09Z|
|Branch|https://github.com/twisted/twisted/tree/irc-tags-6667-4|

This patch intends to enhance the sending of messages sent using sendMessage (twisted.words.protocols.irc.IRC.sendMessage) to allow sending message tags with the message (as specified in the [IRCv3 spec](http://ircv3.atheme.org/specification/message-tags-3.2)).

In order to fully support message tags (which serve as the basis for other 3.2 features as well), we'll need to update the API in a way that maintains compatibility (a new set of methods?).  This patch also does not change IRCClient, mostly because I'm much less familiar with its use.

Even with these drawbacks, this will at least allow servers to support tags which only originate from the server (such as [server-time](http://ircv3.atheme.org/extensions/server-time-3.2)).

Attachments:

* [irc-tags.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/0e1d8cfe9e907cab976b6a2620482ad28e5e72f0.diff) (2532 bytes) - added by ElementalAlchemist on 2013-08-06 03:03:21Z - 
* [irc-tags-with-topfile.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/c4bdbeed6be0f5706bd145a6ead77d17cdb47ef7.diff) (2532 bytes) - added by ElementalAlchemist on 2013-08-08 01:09:16Z - 
* [irc-tags-with-topfile-actually.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/a6be644421655f7dbec9d14bad83d7afd109494b.diff) (2937 bytes) - added by ElementalAlchemist on 2013-08-08 01:11:53Z - 
* [patch-6667-part-2-on-trunk.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/2fea92482f10e7690725b435bf2fa3c58c8ad051.diff) (2933 bytes) - added by ElementalAlchemist on 2013-10-12 17:57:01Z - 
* [patch-6667-part-2-on-branch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/c21889a86623399f33e58fe367572e559382420e.diff) (1021 bytes) - added by ElementalAlchemist on 2013-10-12 17:57:32Z - 
* [patch-6667-part-2-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/e7b02a84aa7ce7aa738f12ac5ea12fe0e65cb864.diff) (3382 bytes) - added by ElementalAlchemist on 2013-10-12 18:29:05Z - 
* [patch-6667-part-2-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/f40772547b4dac2d81ac93bbff2cec6def3e6f28.diff) (3901 bytes) - added by ElementalAlchemist on 2013-11-27 00:55:23Z - 
* [patch-6667-part-3-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/aff5f095a85b0cdb86339aec854caf09e60e00b3.diff) (9110 bytes) - added by ElementalAlchemist on 2014-02-18 04:53:16Z - Next patch based off of the new branch
* [twisted-dev-log.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/47b10e021593470de29bff6c6048d60a25aebbe0.txt) (3897 bytes) - added by ElementalAlchemist on 2014-02-19 03:49:32Z - We ended up having a very short conversation.  I upload the log here for posterity (even though the implementation will be in a future ticket).
* [patch-6667-part-3-doc.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/06d7404a369dd54e10f29b8be4f82948dacd7ea8.diff) (5800 bytes) - added by ElementalAlchemist on 2014-02-27 01:18:09Z - 
* [twisted-6667-redo1-patch1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/78852a8cb7e2f7ebdd03f4a471cfa91219265076.diff) (7886 bytes) - added by ElementalAlchemist on 2014-04-27 22:30:22Z - Completely redone update of this patch.  Based on trunk.
* [twisted-6667-redo1-patch2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/d0ba5248804da43e8052bbdd698160e6a41eced6.diff) (16032 bytes) - added by ElementalAlchemist on 2014-05-03 20:40:39Z - 
* [twisted-6667-redo1-patch3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/47abf15873a54701033234bf21c6c993a51c6886.diff) (20928 bytes) - added by ElementalAlchemist on 2014-05-05 02:12:52Z - 
* [twisted-6667-redo1-patch4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/8ba1017d2a3b5bcb48797fe39316044a0e2c5003.diff) (20888 bytes) - added by ElementalAlchemist on 2014-05-25 15:44:33Z - 
* [twisted-6667-redo1-branch3-update1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/dd5f8e55b1f20afd9d3dff43195ea6e57ef54800.diff) (13169 bytes) - added by ElementalAlchemist on 2015-04-27 02:47:11Z - First update patch based on irc-tags-6667-3 branch
* [twisted-6667-branch4-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/04395d7409d75043ca3ac9f5e58289d5ac8d25e5.diff) (17111 bytes) - added by ElementalAlchemist on 2015-06-20 18:54:00Z - 
* [twisted-6667-branch4-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76f/76f2078abcddac362c694aff456089f0ae679124/f5304f253bfd3bcef82f04229dc9dcef307a152c.diff) (3722 bytes) - added by ElementalAlchemist on 2015-08-18 04:45:16Z - Update based on latest branch commit

<details><summary>Searchable metadata</summary>

```
trac-id__6667 6667
type__enhancement enhancement
reporter__ElementalAlchemist ElementalAlchemist
priority__normal normal
milestone__ 
branch__branches_irc_tags_6667_4 branches/irc-tags-6667-4
branch_author__adiroiban__glyph__rwall__habnabit adiroiban, glyph, rwall, habnabit
status__closed closed
resolution__fixed fixed
component__words words
keywords__None None
time__1375758129000000 1375758129000000
changetime__1445127023994533 1445127023994533
version__None None
owner__glyph glyph
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_4"></a>|@habnabit commented|
|-|-|

(In [40273]) Branching to 'irc-tags-6667'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_5"></a>|@habnabit commented|
|-|-|

(In [40274]) Apply patch from ElementalAlchemist; refs [#6667](https://github.com/twisted/twisted/issues/6667)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_6"></a>|@habnabit set owner to ElementalAlchemist|
|-|-|

Hi! Thanks for the patch! Sorry about the long wait before the review.

[Your buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/irc-tags-6667) looks good, except with the minor issue of you have trailing whitespace in your patch.

Just one issue—what should happen if the user passes something that's not a 1- or 2-tuple? Right now, your code will silently pass these values through, which doesn't seem like ideal behavior to me. Maybe do two loops through the tag list, and first ensure that everything has length 1 or 2?

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ElementalAlchemist commented|
|-|-|

Awesome!

The trailing space appears to be because I left the default indent on a blank line.  Whoops.

Originally I was all like, "if they're going to pass something that doesn't make sense, it doesn't matter if something that doesn't make sense comes out" (it's very easy to correctly parse tags, so something that's like, "test=what=is=this=jerks" shouldn't break anything but probably isn't really what they want).  Truncating the list there is at least just as good, so I've gone ahead and done that.

Also: I was hoping to be able to discuss how to go about allowing the API to receive tags in addition to sending them.  I understand a patch to do that would get its own separate ticket (but this one is 6667! :( ) but I'd like to at least get advice on where to go with it.  (If we should do this on IRC instead, let me know and I'll head in.)

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|ElementalAlchemist commented|
|-|-|

I'm guessing you wanted the next patch based on the branch, but I wasn't entirely sure, so I also included one based on the trunk.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_9"></a>|@habnabit commented|
|-|-|

IRC is a good place to discuss Twisted tickets in general. `#twisted-dev` on freenode is the channel for discussion of development of Twisted.

I still think that an error should be raised instead of silently trying to guess what the user wants. It also probably doesn't handle 0-tuples in a way that's expected by a client consuming tags.

Also, yes, please include future patches based on the branch.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|ElementalAlchemist commented|
|-|-|

Alright, I've just submitted the one based on the branch this time.

This one raises errors for tag problems and adds a few tests to check for those errors.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_12"></a>|@wallrj commented|
|-|-|

(In [41253]) Apply patch-6667-part-2-1.diff by ElementalAlchemist. Refs [#6667](https://github.com/twisted/twisted/issues/6667)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj commented|
|-|-|

(In [41254]) Revert patch-6667-part-2-1.diff by ElementalAlchemist. There's a newer one. Refs [#6667](https://github.com/twisted/twisted/issues/6667)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj commented|
|-|-|

(In [41255]) Apply patch-6667-part-2-2.diff by ElementalAlchemist. Refs [#6667](https://github.com/twisted/twisted/issues/6667)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj commented|
|-|-|

I haven't got time to do a full review, but I thought it would be helpful to apply the latest patch and start a build.

The results will be here shortly:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/irc-tags-6667

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj set owner to ElementalAlchemist|
|-|-|

Thanks ElementalAlchemist.

I'm not particularly familiar with twisted.words or with the IRC protocol, so
lets call this as a cosmetic review.

Notes:
 1. Merges cleanly
 2. Build results:
    1. Mostly green: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/irc-tags-6667

Points:
 1. Minor whitespace issue
    1. https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1689/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
 2. [branches/irc-tags-6667/twisted/words/protocols/irc.py](branches/irc-tags-6667/twisted/words/protocols/irc.py)
    1. sendMessage
       1. It would be better to document the optional 'tags' keyword argument
          using epydoc so that it is rendered consistently in the API docs
          https://twistedmatrix.com/documents/current/api/twisted.words.protocols.irc.IRC.html#sendMessage
          eg
          ```
          @param tags: An optional list of message tags. Must be supplied as a
              keyword argument. One-tuples should be passed for tags whose presence
              has a meaning (i.e. those that do not show a value).
          @type tags: L{list} of 1-L{tuple} or 2-L{tuple}.
          ```
       2. And at the same time add epydoc for the other required and optional
          arguments (command, parameter_list, prefix)
          http://epydoc.sourceforge.net/manual-epytext.html and
          https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto9
 3. [branches/irc-tags-6667/twisted/words/test/test_irc.py](branches/irc-tags-6667/twisted/words/test/test_irc.py)
    1. "285             if 'tags' in prefix: "
       1. Seems to me that this argument validation code would be better broken
          out into a separate private function (eg _validateTags) which can then
          be tested in isolation.
       2. Maybe also have another function for serialising tags into colon
          separated k=v pairs; again, with it's own separate tests.
    2. "1698            self.assertEqual(str(error)"
       1. Maybe the tests would be less brittle if you check the ``message``
          attribute of the error instead of the stringified version of the
          error.
    3. "1721            Passing invalid tags to L{IRC.sendMessage} raises a C{ValueError}. "
       1. All these tests have the same docstring. Please give each test a
          specific docstring.
    4. Add a test for empty tag values
       1. https://buildbot.twistedmatrix.com/builds/twisted-coverage.py/twisted-coverage.py-r41255/twisted_words_protocols_irc.html#n295
 4. Documentation
    1. I can't find any howto or IRC examples that demonstrate how to use
       sendMessage. Would you mind adding a short example? And ideally add some
       brief narrative documentation to go with it? Maybe an example of
       implementing a server that supports the "server-time" tag mentioned in
       the ticket description.
       http://twisted.readthedocs.org/en/latest/projects/words/index.html
    2. The documentation has recently been migrated to Sphinx, but this branch
       doesn't have the new documentation directory so I'll merge it forward for
       you.
 5. "discuss how to go about allowing the API to receive tags in addition to sending them"
    1. In ticket:6667#comment:7 you mentioned this, but I can't find any trace of the conversation.
    2. Please join #twisted-dev, discuss it with habnabit etc and paste your conclusions here.
 6. Please raise tickets for any necessary followup work eg "API to receive
    tags" "tag support in IRCClient"

Thanks again and sorry for the delay. As I said above, IRC isn't really my
thing, so feel free to correct any of my points above. But please answer them
and address the ones that you do agree with. Then resubmit for what will
hopefully be a final review by someone who knows IRC better than me.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj commented|
|-|-|

(In [41687]) Branching to 'irc-tags-6667-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_18"></a>|@wallrj commented|
|-|-|

Ok. I merged forward so that you can use the new Sphinx documentation system.
 * Branch: log:branches/irc-tags-6667-2
 * Build Results: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/irc-tags-6667-2

Please submit any future patches against that branch.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj commented|
|-|-|

A couple of points that I forgot:
 1. [branches/irc-tags-6667-2/twisted/words/topfiles/6667.feature](branches/irc-tags-6667-2/twisted/words/topfiles/6667.feature)
    1. You can probably remove "The function " from the start of the feature
       sentence for better consistency with other features in the changelog.
 2. [branches/irc-tags-6667-2/twisted/words/protocols/irc.py](branches/irc-tags-6667-2/twisted/words/protocols/irc.py)
    1. Maybe add an "@see: U{http://ircv3.atheme.org/specification/message-tags-3.2}" to the sendMessage docstring.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|ElementalAlchemist commented|
|-|-|

I've attached a new patch which I believe should resolve most of the points except the one about documentation.  It's late now, so I'll get that going in the next day or two.  Let me know if you think anything here is set up in an unideal way so that it can be fixed. :)

Re: 5: The conversation hasn't happened yet.  I started getting caught up in things, and then I forgot about it.  I'll try to do that soon as well.

Re: 6: Yeah, I figured new tickets would be better places for them than trying to cram them into this ticket.  By this point, adding more things to this ticket will just result in more work for both of us.  I'll discuss how to approach an API for receiving tags with someone, and I'll try to get a patch in for that (in a new ticket) when I can.  I don't know if I'll touch IRCClient because I don't use it and I'm less familiar with how it works.  If it comes up when discussing the receive API, I'll see what I can do.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|ElementalAlchemist commented|
|-|-|

I added some documentation.  I didn't do an example server because I can't think of a good example that has a small enough scope to be an example.  There's a lot going on in the server.

Something I was thinking about before this change becomes super permanent, though: Would the tag list be better as a dict (with no-value tags having a value of None), or do you think this solution is better than passing a dict?

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|ElementalAlchemist commented|
|-|-|

I just uploaded a new patch to implement this in a slightly different way.  It's based on trunk because I basically redid it (even though I ended up borrowing a fair amount from the branch; for example, I wasn't going to re-rewrite that doc).

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ElementalAlchemist removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to ElementalAlchemist|
|-|-|

Thanks!

  1. Use `L{}` instead of `C{}` when you're presenting the name of an object (such as a type).  Use `C{}` for Python expressions and statements.
  1. Use `bytes` or `unicode` instead of `str` unless you really actually seriously mean "the native string type".
  1. The signature of `sendMessage` (and therefore the API documentation) is a little wonky.  I don't see why `sendMessage` should accept arbitrary keyword arguments at all.  It ignores them all except `prefix` (and now `tags`).  This is a good recipe for an error prone API.  What if I forget and pass `tag` instead?  This is an existing defect but this change compounds the problem.  I suggest introducing a method to replace `sendMessage` and giving it a simpler, more normal signature (`sendMessage` can then be trivially implemented in terms of it and probably eventually deprecated and removed).
  1. The API documentation for `tags` needs more information.  What are keys in the dictionary?  What kind of strings (bytes or unicode) are allowed?
  1. New code should use `b""` syntax for bytes literals, if that's what's meant here (for `"@%s %s"`, for example).
  1. `_stringTags` needs a docstring.  Also `b""` if that's what's meant here.
  1. Same for `_validateTags` and `_escapeTagValue`.
  1. New code should avoid `iteritems` (and `iterkeys` and `itervalues`).
  1. The document linked in the `index.rst` is not actually added by this patch.
  1. The loop in `test_sendMessageWithTags` doesn't solve the problem of dictionary iteration order.  You can't rely on two different dict instances with the same keys and values iterating in the same order.  If you need a particular order in the tests to make it possible to make an assertion then you should have the implementation enforce that order.  For example, sort the tags before emitting them.
  1. The loops containing assertions in `test_stringTags` produce bad failures.  If one of the assertions fails then the person reading the failure has no way to find out which assertion failed.  They also don't know if later iterations of the loop would have had failed assertions.  Try to write tests that only call one assertion method one time.  Some of these test methods may be better written as multiple test methods.
  1. Similar comment for `test_validateEmptyTags` and `test_errorTagsWithInvalidChars`
  1. The docstring for `test_sendMessageWithTags` talks about the "proper" string.  How does a maintainer know what proper is?  Better than saying "proper" here would including  a link to a standard defining what the correct behavior is.
  1. Similar comment for `test_stringTags` and `test_sendMessageWithPrefix`.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_25"></a>|ElementalAlchemist commented|
|-|-|

2,5,6,7. I'm sticking with `str` for now unless there's something I don't know about that requires me to pick one.  Rationale:

- It's easier to work with strings than with picking a type of either `bytes` or `unicode`, especially if we're planning to maintain compatibility with both Python 2 and 3 for the time being.  The `bytes` type seems more difficult to work with in Python 3 (looking at the lack of string operations an IRC server may be likely to perform, like `.format`).

- I felt silly having to clarify one type for Python 2 and one for Python 3.  Whatever `str` aliases to in their given implementation works fine.

- It doesn't make a difference as to the validity of the IRC message whether we store it in one type or the other, and we're not doing any length checking on it.

If any of you all think that the type makes a difference, I have no problem changing it, but I'm looking at ease-of-use as well as functionality, and I expect users to want to be able to just plug in the type of thing they're most likely already currently using.  Same thing with the fact that I didn't add bytes literals--most users will want to pass a string, and that will raise an exception in Python 3 unless we convert it first.

3. I gave it a name, but I welcome suggestions for better ones if you have them.  I couldn't think of a better one than `sendMessage`, which is already used by the old one, so for now I went with a variation on that, which might be a bit confusing during the transition.

7. Separate from above: The escaped value requirements are pending addition to the spec (https://github.com/ircv3/ircv3-specifications/pull/78).  If this patch sits pending to merge until the standard is updated based on my addition of the escaping, that's fine.  This is here so that you know from where I'm getting that, since it's perhaps a bit important and not yet at the link in the comments (though it will be).

9. I suck at svn.  It's added now.

Other than either the objections or the points raised above, I think I addressed all the issues from this round, but please also verify that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

Thanks for your continued efforts on this!

> It's easier to work with strings than with picking a type of either bytes or unicode

This is an interesting argument.  I think the trouble is that while it sounds compelling it's not quite true.  It's only easier until you actually try to make the code run on Python 3 and then everything falls apart.  Perhaps that means it's not even worth trying to write Python 3 friendly code here: all of the Python 3 porting effort will just wait until someone *actually* tries to port the code.

However, this still means you should probably document everything as `bytes` instead of as `str`.  Saying the type is `str` without trying to demonstrate it actually works on Python 3 is just misleading.

The part where just using `str` and hoping it works on Python 3 gets problematic is the part where the low-level implementation can actually only work with `bytes`.  Since this code only runs on Python 2 right now and `str` is `bytes` it all *seems* to work if you just use the native string type.  This breaks as soon as the native string type is unicode, as it is on Python 3, though, because you can't put unicode into a socket.

(All of which sort of undermines my comment about not even trying to write Python 3 friendly code :/)

> I felt silly having to clarify one type for Python 2 and one for Python 3. Whatever str aliases to in their given implementation works fine.

I don't understand this comment.  Perhaps my review comment about `bytes` vs `unicode` wasn't clearly stated.  My suggestion was that you have to pick *either* `bytes` *or* `unicode` and then document that.  Not that you should document it as `bytes` on Python 2 and `unicode` on Python 3 - that's exactly what `str` means and indeed it would be pointless to expand the documentation to be more verbose but still mean exactly the same thing. :)  The purpose of selecting one or the other of these two types is that the implementation will actually only work with one or the other.  Unless you go out of your way to make it work with both.  In which case the documentation should actually say "bytes or unicode" - regardless of Python version, because if the code works with bytes and unicode then it works with bytes and unicode and what `str` means is basically irrelevant.

> It doesn't make a difference as to the validity of the IRC message whether we store it in one type or the other, and we're not doing any length checking on it.

I don't understand this comment either.  I don't understand why a length check would make any difference here.

Whether the type is `bytes` or `unicode` may not make a difference to the validity of the IRC message but it does make a difference to whether you can *send* the message or not.  If it is unicode then you cannot send it because you can only write bytes to sockets.

> If any of you all think that the type makes a difference, I have no problem changing it, but I'm looking at ease-of-use as well as functionality, and I expect users to want to be able to just plug in the type of thing they're most likely already currently using. 

Unfortunately it is quite likely that the "ease of use" ship has already sailed (Python 3 made the decisions it made, we can't do anything about that).  If anything then I suspect the best choice to make here with respect to ease of use is to pick `unicode` as the type and do all of the necessary encoding and decoding to make that work.  `unicode` is the most correct type to use to represent text and I gather the information being represented here is text.  The way the specification outlines handling of non-ascii may have some bearing here, though.  For example, what is the correct way to handle the tag `☃=bar`?  Does it make a difference if it is part of a vendor prefix?  For example, are the rules for `☃.com/foo=bar` any different?  Does punycode or idna get involved here?  A quick skim of the spec doesn't reveal the answers to these questions to me (but it was very quick, I could easily have missed something).

Python 2 developers are accustomed to typing u"..." and they have `from __future__ import unicode_literals` at their disposal so I'm not too worried about the usability burden to them of having to use the correct type to represent text.

> Same thing with the fact that I didn't add bytes literals

Using bytes literals is just about being explicit in the implementation.  It says "this is bytes" instead of "this is either bytes or unicode".  This is basically necessary for the same reason as I gave above about the implementation actually working on Python 3.  If the messages turn into unicode then you can no longer put them into sockets.  Of course, if it turns out the correct type for some of these objects (for example, the tag names and values) is unicode then all of the code that represents such objects using literals should use unicode literals instead of bytes literals.

>  I gave it a name, but I welcome suggestions for better ones if you have them. I couldn't think of a better one than sendMessage, which is already used by the old one, so for now I went with a variation on that, which might be a bit confusing during the transition.

An idea might be `sendTaggedMessage` - perhaps coupled with the idea that maybe `sendMessage` won't even offer this feature (that cheat makes it seem less weird that there's both a `sendMessage` and a `sendTaggedMessage`).  I won't claim this is a brilliant (or even good) idea, though.

A different line of thinking could perhaps shortening the name to `send` (what is ever sent apart from "messages" - and what does "message" even mean?  basically nothing) or picking a synonym (this is probably the worst idea) like `sendCommand`.

And here's an idea I do somewhat like, a new `Message` class which represents all of the state of a message and is the sole argument to a new `send` method.  This is bloating scope a little bit though so I won't propose it as a necessary change here.

Sorry about all this bytes/unicode thrashing.  I wish Twisted had better documentation available explaining how we're trying to deal with this issue. :(  Thanks again for working on this.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_27"></a>|ElementalAlchemist commented|
|-|-|

> This breaks as soon as the native string type is unicode, as it is on Python 3, though, because you can't put unicode into a socket.

Blargh, Python 3.

This seems to be against the whole "Unicode for everything!" for which they seemed to be going by making unicode the default type, but then I understand that that's not our decision to make and that that ship has long sailed, but still: rant.

I'll look into either converting things to bytes or trying to make it dual-compliant (maybe compile a `unicode` string and then convert to `bytes` for `sendLine`).  Other than joining strings together, we aren't (shouldn't be) doing any string manipulation, so it's probably OK to convert everything to `bytes` from the start once we get it.

> I don't understand this comment either. I don't understand why a length check would make any difference here.

Line length limits in the IRC standard are defined in bytes.  Sending a line of 512 unicode characters isn't necessarily valid, but the length of the `bytes`-string is a correct measurement.  We don't currently have this check, but if we do something like that in the future, how we count that will be relevant.

> The way the specification outlines handling of non-ascii may have some bearing here, though. For example, what is the correct way to handle the tag ☃=bar? Does it make a difference if it is part of a vendor prefix? For example, are the rules for ☃.com/foo=bar any different? Does punycode or idna get involved here? A quick skim of the spec doesn't reveal the answers to these questions to me (but it was very quick, I could easily have missed something).

This one isn't your fault: apparently the IRCv3 working group is awful at keeping the website updated.  (They really need to get on that.)  This is one that's [already part of the spec](https://github.com/ircv3/ircv3-specifications/blob/master/specification/message-tags-3.2): punycode is used for non-ASCII hosts.  (You'll notice the current validation code would reject your example.  I didn't think we needed to do the punycode translation for them.)

> Python 2 developers are accustomed to typing u"..." and they have from __future__ import unicode_literals at their disposal so I'm not too worried about the usability burden to them of having to use the correct type to represent text.

I'm not, but I've been yelled at for passing unicode to sendMessage before.  That said, it's probably simple enough to make things into Unicode strings (though I'd prefer to err on the side of not breaking current code that doesn't use Unicode strings by requiring them).

> An idea might be sendTaggedMessage - perhaps coupled with the idea that maybe sendMessage won't even offer this feature (that cheat makes it seem less weird that there's both a sendMessage and a sendTaggedMessage). I won't claim this is a brilliant (or even good) idea, though. 

> A different line of thinking could perhaps shortening the name to send (what is ever sent apart from "messages" - and what does "message" even mean? basically nothing) or picking a synonym (this is probably the worst idea) like sendCommand.

If I saw a message called `send`, I'd initially assume it was a synonym of `sendLine`.  I think we need at least some indication of "send this as an IRC line" as opposed to "compile this data into an IRC line and send it."  That said, I'll look at those other ideas and ponder them for a while.

> Sorry about all this bytes/unicode thrashing. I wish Twisted had better documentation available explaining how we're trying to deal with this issue. :(

I understand.  It's a discussion we need to have apparently, especially with my lack of experience with Python 3 (so far, anyway--I imagined it as less of a departure from Python 2 as it is, but I'm starting to see how all the people who think Python 3 is too different are justified).  I'll look into what I can do and come back with hopefully something.

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_28"></a>|ElementalAlchemist commented|
|-|-|

I think I got the bytes/unicode thing going.  Please check for sanity and any other issues.

I've punted on renaming the method for now while I ponder the best route, and I'm punting on updating the supplied documentation until I decide on a method name, but I'll try to get that done before this gets merged.

I've added a set of test cases that work like the ones for sendMessage but don't call the sendMessage function (since it's now a wrapper).  When sendMessage gets removed in the future we should just be able to remove the sendMessage test cases and keep the ones for the new function.  (If it's not removed, it's probably a good idea to keep all of them.)

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_29"></a>|ElementalAlchemist commented|
|-|-|

Updated patch: The final accepted version of the value defaulting specification omits the replacement of the BEL character.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun removed owner|
|-|-|

Last submitted for review by ElementalAlchemist so probably not being reviewed by ElementalAlchemist.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_31"></a>|@glyph commented|
|-|-|

(In [43724]) Branching to irc-tags-6667-3.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_32"></a>|@glyph commented|
|-|-|

Uploaded to the buildbot and builds forced.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_33"></a>|@glyph set owner to ElementalAlchemist|
|-|-|

Hi ElementalAlchemist,

Thanks again very much for your contributions.  Sorry this review took so terribly long; we do not have nearly enough code reviewers.  I hope you're still interested in this, and perhaps we can get you on board as a reviewer as well, so others don't need to wait so long either :).

This is actually looking very close, but there are some minor issues that need addressing.

  1. [the buildbot](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/irc-tags-6667-3) found some [coding standard compliance issues](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2470/steps/run-twistedchecker/logs/new%20twistedchecker%20errors) and an [undefined name](https://buildbot.twistedmatrix.com/builders/pyflakes/builds/2130/steps/pyflakes/logs/new%20pyflakes%20errors).
  2. The documentation appears to be cargo-culting the terrible formatting created by our automatic lore to sphinx conversion tool.  Please collapse all the redundant newlines in `ircserverclientcomm.rst`.  Also, apparently this isn't documented in our doc writing standard anywhere, but we use what we call "semantic newlines" in our Sphinx documentation - that is, only put a newline after a complete sentence.  (This is intended to reduce the churn produced in diffs from re-wrapping every paragraph.)  Note that this applies only to ReST documentation; epytext in python files is still wrapped to 79 characters.

As I was perusing the code, I also noticed these things that you might want to address if you so desire;

  1. you have several IRCv3 specification links.  In epytext, you can use `U{link name <https://link text>}` syntax.
  2. please use `https` links everywhere that the target site supports them.

If you could address the first two issues, and then address the optional comment to your satisfaction, I think we can land this.  Once again, I'm sorry it took so long.  Thanks, and I look forward to reviewing it a final time!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_34"></a>|ElementalAlchemist removed owner|
|-|-|

I made the changes, and hopefully the documentation is now up to spec. The IRCv3 site doesn't seem to support HTTPS, but I updated the links to the RFC to use HTTPS.

I made a couple additional changes in this diff, including one change for a last-minute tweak before IRCv3.2 was finalized and updating all the IRCv3 links since the site seems to have moved.

There was one buildbot comment I didn't get because it was complaining about some comment formatting at the beginning of a comment, but place it seemed to direct me to was in the middle of a docstring. If that's still a problem, please provide some guidance on how to parse buildbot output. :P

Although I made reference to and originally also planned to add (in a separate ticket/patch, of course) the other side (i.e. receiving messages with tags), I've since replaced the Twisted Words IRC class in my application with my own IRC parser built on LineOnlyReceiver, so I no longer have any use for it to be here. Hopefully someone will be enterprising enough to take on this task for Twisted.

Attached below is an update to the branch that appears to have been made with the latest changes.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_35"></a>|@adiroiban commented|
|-|-|

(In [45017]) Branching to irc-tags-6667-4.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_36"></a>|@adiroiban set owner to ElementalAlchemist|
|-|-|

Thanks for the latest changes.

I have apply your changes and sent them to buildbot... in a separate branch.

At leas one test is failing.

Since this review is taking so long I am trying to add my comments.

Please not that this is a high level review and I have no idea how IRC protocol works and how tags are used.

But I hope that we can fixe some problems so that future reviewers will have to review and commend on fewer problems.

Maybe I am also repeating some problems which were already discuss... sorry for that.

-------

documentation has a few rst/sphinx markup errors.

you can test the documentation using

```
cd docs
make html
OPEN _build/html/words/howto/ircserverclientcomm.html
```

here are the errors

```
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:29: ERROR: Error in "code-block" directive:
maximum 1 argument(s) allowed, 5 supplied.

.. code-block:: python
    user.sendMessage("COMMAND", param1, param2, prefix=server.name)
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:31: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:36: ERROR: Unexpected indentation.
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:48: ERROR: Error in "code-block" directive:
maximum 1 argument(s) allowed, 26 supplied.

.. code-block:: python
    channel.pastMessages = [
        ("I sent some text!", "author!ident@host", datetime object representing the when the message was sent),
        ("I did, too!", "someone-else!ident@host", another datetime object)
    ]
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:53: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:57: ERROR: Error in "code-block" directive:
maximum 1 argument(s) allowed, 4 supplied.

.. code-block:: python
    sendingTags["server-time"] = "{}Z".format(message[2].isoformat()[:-3])
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:59: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:63: ERROR: Error in "code-block" directive:
maximum 1 argument(s) allowed, 6 supplied.

.. code-block:: python
    user.sendMessage("PRIVMSG", user.nickname, message[0], prefix=message[1], tags=sendingTags)
/home/adi/chevah/twisted/docs/words/howto/ircserverclientcomm.rst:87: WARNING: Explicit markup ends without a blank line; unexpected unindent.

```

--------

`yourself.  The rest` ... double space

Instead of

```
This works as the protocol for your Factory class, and it will 
```

maybe


```
This works as the protocol for your Factory class.
It will ...
```

------

see the error

```
syntax for sending for sending messages
```

-------

I think that we should deprecated sendMessage ... this can be done in a separate ticket... but there is no need to update the old sendMessage method to support tags.

So instead of a `This method remains for backward compatibility.` just put a depreacation warning.

I don't like sendMsg name as it is an abreviation. Maybe we can rename it as sendCommand as the code is actually sending an IRC command.	

----

In other parts of twisted I see that docstring has `@return` first, followed by `@rtype`

-----

For test_sendUnicodeMessage it would be nice to also put some non-ascii values into the param and check that by default they are encoded to UTF-8

Docstring for test_sendUnicodeMessage looks like a copy paste from another test ... I think that it should be updated to betted describe the fact that unicode are accepted and automatically encoded to utf-8

----

test_sendUnicodeMessageNoCommand does this test has anything to do with Unicode values... or is about None value? ...

maybe it is a copy/paste error and should be named test_sendMsgNoCommand ... similar for the other tests.

--------

_stringTags, _validateTags and _escapeTagValue are just implementation details . You can leave these tests but you should also write integration tests to check that sendMsg does tags validation/escaping.


--------

You should avoid inline comments and whe using inline comments use the PEP8 formating.
Comments shoudl be full senteces.


Not sure if this comment `# Make sure the tags we got are sane` adds any new info to the test.

------ 

I think that the docstring of `def irc_unknown` from the documentation should explain what the fuction does.



-------

My suggestion is to leave sendMessage method untouched and create a separate ticket to deprecate it.

Create a new sendCommand method.

Update documentation to explain how to send messages with tags.

--------


I don't like the way self.encoding is handled ... I see a lot of code duplicated conditionals around this member. I would argue that since utf-8 is safe for irc, we should initialize the encoding with UTF-8 rather than with None.

--------

Please see my comments and come back with a new patch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_37"></a>|ElementalAlchemist removed owner|
|-|-|

Failing test: Whoops! Forgot to update it.

Looks like I got a bit overzealous on condensing my doc formatting. Fixed that. Thanks for the advice on how to see doc errors! :)

I changed `sendMsg` to `sendCommand`. I don't think it's perfect--messages coming from the server are not always going to be commands (they're commonly numerics)--but it's the best I have at the moment.

I added the deprecation. I left the tags support in it, though. I insisted on keeping the original `sendMessage` since I'd actually taken advantage of it having keyword arguments in my program by adding some more keyword arguments to make it easier to route messages as-is to both local and remote users. I suppose that now that I neither route messages that way nor use Twisted Words in the same capacity, it could be removed, but at this point it's already there. We can avoid adding future enhancements, but I have no problem with this one. (If there are strong objections to this, I can remove tags support from `sendMessage`.)
Whatever we do, I think we should do something with `sendMessage` here so there's not a (even temporary) disparity between what `sendMessage` can do and what its effective replacement can do. I'd prefer we do one or both of
* mark the function as deprecated
* add tags support (but, if it's deprecated, maybe not future enhancements)

`test_stringTags` already had a version of it that worked with `sendCommand`. I added integration tests for the other functions as well. I also renamed a bunch of tests.

Does `irc_unknown`'s docstring not explain what it does? I'm not sure what to add to it.

This patch has been in the works for two years now, and I'm, quite frankly, tired of it. (Especially since it's now no longer particularly important to me.) If we want to change how the `encoding` property is handled, let's do it separately.

Patch incoming after the post.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_38"></a>|@adiroiban commented|
|-|-|

Many thanks for working on this!

I have applied your latest patch and sent it to buildbot.

The deprecation of sendMessage should be done in a separate ticket and should
come with dedicated tests. If we want to keep it in this ticket it needs tests
and a new .removal topfile to advertise the changes in the changelog.

If you no longer need this feature I think that there is no reason for hurry
to merge the changes.

I am sorry that this patch took so long ... there are few Twisted reviewers and I feel that there is no maintainer for twisted.words.irc.

-------

I have pushed my changes which removes the deprecations of sendMessage and any changes related to sendMessage ... once sendCommand is merged we can deprecated
sendMessage in a separate ticket.

I have removed the duplicated tests for private methods.

I have renamed the public argument from `params` to `parameters`


-----------


I think that the patch is ok, but I am doing just a high level review.

It would help if someone with a good understanding of IRC to do a final review of the
the new API.

If this patch will stay un-reviewed I will merge it in a few days.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_39"></a>|@glyph set owner to ElementalAlchemist|
|-|-|

Thanks for your sustained interest in this, ElementalAlchemist.  Sorry it seems to be getting bogged down.

I wish I could say that I would just merge it, but I still see a few issues :).

  1. All of the documentation documents `sendMessage`, but it looks like the new, good thing here is actually `sendCommand`.  If the intent is to replace `sendMessage`, then please 
  2. Other than its direct unit tests, nothing validates that `sendMessage` actually works.  Replacing _every_ caller of `sendMessage` or `sendLine` with `sendCommand` might make the patch too big to be reasonably reviewed, but it seems to me that replacing at least _one_ caller (say, the `privmsg` method?) would be a good start, or perhaps choose a method or two within `twisted.words.service`.
  3. Nothing in `sendCommand`'s docstring really explains what its purpose is relative to `sendMessage` either.  Which should be preferred?  Why?  Why do two exist?
  4. Before closing this ticket, we need to ensure that there is a ticket to remove old users of `sendMessage` _within_ Twisted, and another ticket to do the deprecation once that is done; we don't want to leave the old code around any longer than we have to.

Basically I think that the code looks good, the test coverage looks good, but introducing this as-is will just produce even more confusion for developers consulting this class's API, so the story in the documentation needs to make it clear when and why you'd use the new method over the old.  Ideally we'd have one internal use case to validate the new method.

This ticket has been outstanding for way too long, and you've done everything we've asked (although previous reviews have missed important stuff), so please feel free to ping me personally to ensure that you get a timely review for the next round :).

**Comment by twisted-trac** [comment]
|<img alt="ElementalAlchemist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_40"></a>|ElementalAlchemist removed owner|
|-|-|

1. Whoops! I wrote that doc back when this tickets was "add a new keyword argument to `sendMessage`," and I never remembered to update it. Fixed.
2. I don't use those functions, so I don't think about them. :P They seem to be good candidates for replacement (they all use `sendLine`, but to send lines that could just as easily be sent with `sendCommand`), so I'll make a couple of them use `sendCommand`.
3. I made a go at preferring/not-preferring one in a docstring. I forget how to link to other methods (or if there's a way to do that). Let me know if we need that (and remind me how to do it).
4. Created [#8000](https://github.com/twisted/twisted/issues/8000) and [#8001](https://github.com/twisted/twisted/issues/8001).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_41"></a>|@glyph commented|
|-|-|

The branch is updated and the builders built.  The test failure appears unrelated to the code here, since it's present on trunk; I'll write to the mailing list about that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_42"></a>|@glyph set owner to @glyph|
|-|-|

At long, long last:

L G T M

Will merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_43"></a>|@glyph set status to `closed`|
|-|-|

(In [45820]) Merge irc-tags-6667-4: Add tags to sendMessage.

Author: ElementalAlchemist

Reviewer: glyph, habnabit, exarkun, rwall, adiroiban

Fixes: [#6667](https://github.com/twisted/twisted/issues/6667)

twisted.words.protocol.irc.IRC now has a sendCommand() method which can
send messages with IRCv3-style tags.
